## Introduction to Python
## JN1

## Basic Jupyter

In [4]:
#code line
print
print('Welcome to GISC405')

Welcome to GISC405


In [2]:
print?

Docstring:
print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)

Prints the values to a stream, or to sys.stdout by default.
Optional keyword arguments:
file:  a file-like object (stream); defaults to the current sys.stdout.
sep:   string inserted between values, default a space.
end:   string appended after the last value, default a newline.
flush: whether to forcibly flush the stream.
Type:      builtin_function_or_method


In [3]:
###use show contextual help
print("Welcome to GISC405")

Welcome to GISC405


# Section 1. Basic Python

### Section 1.1. Basic Arithmetic Operations

In [156]:
a = 3

In [157]:
a

3

In [158]:
b = 2

In [159]:
a**b

9

In [160]:
a*b

6

In [161]:
c = a**b
d = a*b

In [162]:
print('square of',a,'is',c)
print(a,'multipled by',b, "is",d )

square of 3 is 9
3 multipled by 2 is 6


In [163]:
5//4

1

In [164]:
5/4

1.25

In [165]:
int(5/4)

1

In [166]:
int?

Init signature: int(self, /, *args, **kwargs)
Docstring:     
int([x]) -> integer
int(x, base=10) -> integer

Convert a number or string to an integer, or return 0 if no arguments
are given.  If x is a number, return x.__int__().  For floating point
numbers, this truncates towards zero.

If x is not a number or if base is given, then x must be a string,
bytes, or bytearray instance representing an integer literal in the
given base.  The literal can be preceded by '+' or '-' and be surrounded
by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
Base 0 means to interpret the base from the string as an integer literal.
>>> int('0b100', base=0)
4
Type:           type
Subclasses:     bool, IntEnum, IntFlag, _NamedIntConstant


In [167]:
5.5//4 # Can someone tell me what this oepration does?

1.0

In [168]:
7%4 # Can someone tell me what this operation does?

3

In [169]:
7//4

1

In [13]:
a = list([1,[2,3]]) # A list in Python is a collectoon of different "objects" or variables with different data types

In [14]:
a[1]

[2, 3]

In [15]:
a[1]

[2, 3]

In [173]:
a = list(['kit','marwan', 'sky'])

In [174]:
a[2]

'sky'

In [175]:
len(a)  # len, short for lenght, returns the number of items in the list

3

In [176]:
len("12345")

5

In [4]:
b = dict([('kit', 5), ('sky', 'newstudent'), ('marwan', -5)])  # A dictionary stores values into specific variable names that can be called directly form the list

In [178]:
b['sky']

'newstudent'

In [7]:
# Dictionaries can also be set using braces {} without invoking the function "dict"
MetData={'Temperature':[20,15,18],'Wind Speed':[3,6,10]}

In [180]:
MetData['Temperature'] # calling a specific part of the dictionary

[20, 15, 18]

In [ ]:
b['x']  # calling a specific part of the dictionary

### Section 1.2. Conditional Statements

In [27]:
a = 4
b = 2

In [20]:
if (a > 3):
    e = a
elif (a < 3):
    e = 3
else:
    e = 0

In [21]:
e

0

In [28]:
if (a > 3) and (b <=2):
    e = a + b
else:
    print('try again')

In [29]:
e

6

Other conditional statements include switch/case pairs and won't be discussed here since we won't use it for this course.

### Section 1.3. Loop Statements

In [10]:
a = 0

In [11]:
r=range(0,10)
list(r)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [12]:
for i in range(0,10):  # looping the index i from 0 to 10 in steps of 1
    a = a + i

In [13]:
list(range(0,10))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [14]:
range?

Init signature: range(self, /, *args, **kwargs)
Docstring:     
range(stop) -> range object
range(start, stop[, step]) -> range object

Return an object that produces a sequence of integers from start (inclusive)
to stop (exclusive) by step.  range(i, j) produces i, i+1, i+2, ..., j-1.
start defaults to 0, and stop is omitted!  range(4) produces 0, 1, 2, 3.
These are exactly the valid indices for a list of 4 elements.
When step is given, it specifies the increment (or decrement).
Type:           type
Subclasses:     


In [15]:
a

45

In [40]:
b = 0

In [41]:
i = 0
while i < 10:
    b = b+i
    i=i+1

In [42]:
b

45

### Section 1.4. Defining a Function

In [5]:
def add_two(a,b):
    c = a + b
    return c

In [6]:
c=add_two(10,20)

In [7]:
c

30

In [8]:
a=[3,4,5]
b=[10,20,30]
#c=a[0]+b[0]
c=add_two(a,b)

In [9]:
# Adding lists is not the same as matricx addition. To do that one needs to define a numpy "array"
import numpy as np
a=np.array([3,4,5])
b=np.array([10,20,30])
d=a+b
d

array([13, 24, 35])

### Section 1.5. Task
Write a function (gisc_exponential) to calculate the exponential of a variable with the base of Euler’s number. Return 0 if the result is smaller than 0.1 else returns the true value

Note: check your python reference card on Larn resources to find the exponential function

In [63]:
# Write your function here


In [17]:
a = 4

In [54]:
def gisc_exponential(x):
    e=2.71828**x
if (c < 0.1):
    e = 0
else:

SyntaxError: unexpected EOF while parsing (3714021828.py, line 5)

# Section 2. Matrix Manipulation
Numpy is "The fundamental package for scientific computing with Python". (https://numpy.org)

### Section 2.1. Importing Libraries

In [19]:
### Libraries are existing packages of functions/classes written by you or someone else. You can utilize the existing functions to speed up your work.

In [20]:
import numpy as np # import all the library

In [21]:
from numpy import sqrt as npsqrt # import a specific operation

In [22]:
np.sqrt(10) # personally I like this way...

3.1622776601683795

In [23]:
np.sqrt(16)

4.0

In [ ]:
np.

### Section 2.2. Creating a Matrix

In [27]:
a = np.arange(0,1000,1)

In [28]:
a

array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
        26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,
        39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,
        52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,
        65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,
        78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,
        91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103,
       104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116,
       117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129,
       130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142,
       143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155,
       156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168,
       169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 18

In [29]:
np.sqrt(a)

array([ 0.        ,  1.        ,  1.41421356,  1.73205081,  2.        ,
        2.23606798,  2.44948974,  2.64575131,  2.82842712,  3.        ,
        3.16227766,  3.31662479,  3.46410162,  3.60555128,  3.74165739,
        3.87298335,  4.        ,  4.12310563,  4.24264069,  4.35889894,
        4.47213595,  4.58257569,  4.69041576,  4.79583152,  4.89897949,
        5.        ,  5.09901951,  5.19615242,  5.29150262,  5.38516481,
        5.47722558,  5.56776436,  5.65685425,  5.74456265,  5.83095189,
        5.91607978,  6.        ,  6.08276253,  6.164414  ,  6.244998  ,
        6.32455532,  6.40312424,  6.4807407 ,  6.55743852,  6.63324958,
        6.70820393,  6.78232998,  6.8556546 ,  6.92820323,  7.        ,
        7.07106781,  7.14142843,  7.21110255,  7.28010989,  7.34846923,
        7.41619849,  7.48331477,  7.54983444,  7.61577311,  7.68114575,
        7.74596669,  7.81024968,  7.87400787,  7.93725393,  8.        ,
        8.06225775,  8.1240384 ,  8.18535277,  8.24621125,  8.30

In [30]:
a.shape

(1000,)

In [38]:
b=np.reshape(a,(10,-1))

In [39]:
b.shape

(10, 100)

In [40]:
b

array([[  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
         13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
         26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,
         39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,
         52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,
         65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,
         78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,
         91,  92,  93,  94,  95,  96,  97,  98,  99],
       [100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112,
        113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125,
        126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138,
        139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151,
        152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164,
        165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 17

In [41]:
c = np.zeros((10,5))

In [42]:
c

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [43]:
D=[[1, 2, 3, 4],[10, 20, 30, 40],[50, 60, 70, 80]]  # matrices can deifned by direct assignment

In [44]:
np.shape(D)

(3, 4)

### Section 2.3. Matrix Operations

In [45]:
d = np.arange(0.0, 1000.0, 1.0).reshape(10,-1) # nesting operations in one line is common in this course

In [46]:
d.shape

(10, 100)

In [ ]:
d

In [47]:
d.dtype

dtype('float64')

In [48]:
b.dtype

dtype('int64')

In [49]:
b.shape

(10, 100)

In [50]:
d.shape

(10, 100)

In [51]:
np.multiply(b,d) # Linear algebra rule for mulitplying

array([[0.00000e+00, 1.00000e+00, 4.00000e+00, 9.00000e+00, 1.60000e+01,
        2.50000e+01, 3.60000e+01, 4.90000e+01, 6.40000e+01, 8.10000e+01,
        1.00000e+02, 1.21000e+02, 1.44000e+02, 1.69000e+02, 1.96000e+02,
        2.25000e+02, 2.56000e+02, 2.89000e+02, 3.24000e+02, 3.61000e+02,
        4.00000e+02, 4.41000e+02, 4.84000e+02, 5.29000e+02, 5.76000e+02,
        6.25000e+02, 6.76000e+02, 7.29000e+02, 7.84000e+02, 8.41000e+02,
        9.00000e+02, 9.61000e+02, 1.02400e+03, 1.08900e+03, 1.15600e+03,
        1.22500e+03, 1.29600e+03, 1.36900e+03, 1.44400e+03, 1.52100e+03,
        1.60000e+03, 1.68100e+03, 1.76400e+03, 1.84900e+03, 1.93600e+03,
        2.02500e+03, 2.11600e+03, 2.20900e+03, 2.30400e+03, 2.40100e+03,
        2.50000e+03, 2.60100e+03, 2.70400e+03, 2.80900e+03, 2.91600e+03,
        3.02500e+03, 3.13600e+03, 3.24900e+03, 3.36400e+03, 3.48100e+03,
        3.60000e+03, 3.72100e+03, 3.84400e+03, 3.96900e+03, 4.09600e+03,
        4.22500e+03, 4.35600e+03, 4.48900e+03, 4.62

In [52]:
b*d # direct one-to-one multiplication, better suted for geospatial data operations, why?

array([[0.00000e+00, 1.00000e+00, 4.00000e+00, 9.00000e+00, 1.60000e+01,
        2.50000e+01, 3.60000e+01, 4.90000e+01, 6.40000e+01, 8.10000e+01,
        1.00000e+02, 1.21000e+02, 1.44000e+02, 1.69000e+02, 1.96000e+02,
        2.25000e+02, 2.56000e+02, 2.89000e+02, 3.24000e+02, 3.61000e+02,
        4.00000e+02, 4.41000e+02, 4.84000e+02, 5.29000e+02, 5.76000e+02,
        6.25000e+02, 6.76000e+02, 7.29000e+02, 7.84000e+02, 8.41000e+02,
        9.00000e+02, 9.61000e+02, 1.02400e+03, 1.08900e+03, 1.15600e+03,
        1.22500e+03, 1.29600e+03, 1.36900e+03, 1.44400e+03, 1.52100e+03,
        1.60000e+03, 1.68100e+03, 1.76400e+03, 1.84900e+03, 1.93600e+03,
        2.02500e+03, 2.11600e+03, 2.20900e+03, 2.30400e+03, 2.40100e+03,
        2.50000e+03, 2.60100e+03, 2.70400e+03, 2.80900e+03, 2.91600e+03,
        3.02500e+03, 3.13600e+03, 3.24900e+03, 3.36400e+03, 3.48100e+03,
        3.60000e+03, 3.72100e+03, 3.84400e+03, 3.96900e+03, 4.09600e+03,
        4.22500e+03, 4.35600e+03, 4.48900e+03, 4.62

In [53]:
np.matmul(b,d.transpose())

array([[  328350.,   823350.,  1318350.,  1813350.,  2308350.,  2803350.,
         3298350.,  3793350.,  4288350.,  4783350.],
       [  823350.,  2318350.,  3813350.,  5308350.,  6803350.,  8298350.,
         9793350., 11288350., 12783350., 14278350.],
       [ 1318350.,  3813350.,  6308350.,  8803350., 11298350., 13793350.,
        16288350., 18783350., 21278350., 23773350.],
       [ 1813350.,  5308350.,  8803350., 12298350., 15793350., 19288350.,
        22783350., 26278350., 29773350., 33268350.],
       [ 2308350.,  6803350., 11298350., 15793350., 20288350., 24783350.,
        29278350., 33773350., 38268350., 42763350.],
       [ 2803350.,  8298350., 13793350., 19288350., 24783350., 30278350.,
        35773350., 41268350., 46763350., 52258350.],
       [ 3298350.,  9793350., 16288350., 22783350., 29278350., 35773350.,
        42268350., 48763350., 55258350., 61753350.],
       [ 3793350., 11288350., 18783350., 26278350., 33773350., 41268350.,
        48763350., 56258350., 6375335

### Section 2.4. Task
Try to use the function you wrote (gisc_exponential) from the above with the numpy array (b), will it work? What is the intepretation of the error?

In [17]:
b=np.arange(0,12,0.5).reshape(12,-1)
np.shape(b)

In [71]:
c=gisc_exponential(b)

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

In [72]:
E=np.exp(b)

In [73]:
E.shape

(12, 2)

In [68]:
E

array([[1.00000000e+00, 1.64872127e+00],
       [2.71828183e+00, 4.48168907e+00],
       [7.38905610e+00, 1.21824940e+01],
       [2.00855369e+01, 3.31154520e+01],
       [5.45981500e+01, 9.00171313e+01],
       [1.48413159e+02, 2.44691932e+02],
       [4.03428793e+02, 6.65141633e+02],
       [1.09663316e+03, 1.80804241e+03],
       [2.98095799e+03, 4.91476884e+03],
       [8.10308393e+03, 1.33597268e+04],
       [2.20264658e+04, 3.63155027e+04],
       [5.98741417e+04, 9.87157710e+04]])

In [29]:
E<10

array([[ True,  True],
       [ True,  True],
       [ True, False],
       [False, False],
       [False, False],
       [False, False],
       [False, False],
       [False, False],
       [False, False],
       [False, False],
       [False, False],
       [False, False]])

In [30]:
np.where(E<10,0,100)

array([[  0,   0],
       [  0,   0],
       [  0, 100],
       [100, 100],
       [100, 100],
       [100, 100],
       [100, 100],
       [100, 100],
       [100, 100],
       [100, 100],
       [100, 100],
       [100, 100]])

In [269]:
b

array([[ 0. ,  0.5],
       [ 1. ,  1.5],
       [ 2. ,  2.5],
       [ 3. ,  3.5],
       [ 4. ,  4.5],
       [ 5. ,  5.5],
       [ 6. ,  6.5],
       [ 7. ,  7.5],
       [ 8. ,  8.5],
       [ 9. ,  9.5],
       [10. , 10.5],
       [11. , 11.5]])

### Section 2.5. Task
Rewrite the gisc_exponential function to work with arrays this time using the np.where and the np.exp function. Call your function gisc_exponential_V2